### ISSUE TITLE
Installation error on Mac

### ISSUE THREAD

**Issue by iamkroot** [iamkroot/trakt-scrobbler#42]
Getting below error. Please help. Thanks heaps mate. 

![image](https://user-images.githubusercontent.com/32594937/72140070-edcdb500-33b5-11ea-83d8-444914205142.png)

_Originally posted by @prabapro in https://github.com/iamkroot/trakt-scrobbler/issues/41#issuecomment-572939036_

**Comment by iamkroot** [comment]
@prabapro Let's do some diagnostics.
`cd` to `/Users/codechilli/Library/trakt-scrobbler` and run `poetry env info`. Could you paste the output here?

**Comment by prabapro** [comment]
> /Users/codechilli/Library/trakt-scrobbler

Hi @iamkroot ,

thanks for helping me further. Here is the output;

**Env info**
```
Virtualenv
Python:         2.7.16
Implementation: CPython
Path:           NA

System
Platform: darwin
OS:       posix
Python:   /System/Library/Frameworks/Python.framework/Versions/2.7
```

**Env Debug**
```
Virtualenv
Python:         2.7.16
Implementation: CPython
Path:           NA

System
Platform: darwin
OS:       posix
Python:   /System/Library/Frameworks/Python.framework/Versions/2.7
Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry debug

Poetry
Version: 1.0.1
Python:  2.7.16

Virtualenv
Python:         2.7.16
Implementation: CPython
Path:           NA

System
Platform: darwin
OS:       posix
Python:   /System/Library/Frameworks/Python.framework/Versions/2.7
```

**Comment by iamkroot** [comment]
> Python:  2.7.16

There's the main problem. It's using python2 as default for some reason.
Try `poetry env remove 2.7`, followed by `poetry env use 3.7`. Then run the `poetry env info` and install script to see if it works.

**Comment by prabapro** [comment]
Hi @iamkroot ,

Thank you for your extended support. I've uninstalled the python3 & poetry and reinstalled both via brew. Now the installation is completed successfully & restarted the machine. Tested with playing videos on both the VLC & PLEX, but nothing is getting updated on Trackt. Is there any other way that I can see whether this one is working successfully? 

Below is the log fyi.

```
2020-01-11 13:34:58,735 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_cache.json doesn't exist.
2020-01-11 13:34:58,735 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_token.json doesn't exist.
2020-01-11 13:34:58,735 - INFO - MainThread - trakt_interface - Access token not found in config. Initiating device authentication.
2020-01-11 13:34:59,579 - INFO - MainThread - trakt_interface - Verification URL: https://trakt.tv/activate
2020-01-11 13:34:59,579 - INFO - MainThread - trakt_interface - User Code: 00D317CF
2020-01-11 13:35:01,021 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-11 13:35:06,678 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-11 13:35:12,557 - INFO - MainThread - trakt_interface - Device auth successful.
```

**Comment by iamkroot** [comment]
@prabapro Good to hear that you were able to get it installed.

Looks like a problem with the background service not starting. That part isn't tested that well on Mac, so you're gonna have to help me here a bit. 

What does `launchctl list | grep com.github.iamkroot.trakt-scrobbler` show?

**Comment by prabapro** [comment]
Hi @iamkroot , sure, it's my pleasure to help you to test this out. Getting nothing for `launchctl list | grep com.github.iamkroot.trakt-scrobbler`. 

![image](https://user-images.githubusercontent.com/32594937/72201961-1de18a80-3480-11ea-9fdd-8818dda25382.png)

However, not sure if this information would help you. I'm using CleanMyMac app & it shows the plist item but disabled. If you want, I can enable it & restart my mac & check.

![image](https://user-images.githubusercontent.com/32594937/72201994-7c0e6d80-3480-11ea-99b4-0827dfa31c20.png)

**Comment by iamkroot** [comment]
Maybe the service is crashing. Try `tail -f /var/log/system.log`.

**Comment by prabapro** [comment]
Hi @iamkroot ,

Here is the log:
![image](https://user-images.githubusercontent.com/32594937/72214084-df4fdc80-3520-11ea-9c8e-66f393983677.png)

From LaunchControl app:
![image](https://user-images.githubusercontent.com/32594937/72214091-10c8a800-3521-11ea-9159-4a7afdde7ecb.png)

**Comment by iamkroot** [comment]
> @prabapro Let's do some diagnostics.
> `cd` to `/Users/codechilli/Library/trakt-scrobbler` and run `poetry env info`. Could you paste the output here?
> 
> Also, paste the output of `poetry debug`. That might help too.

Could you run these commands again and paste the output now?

**Comment by prabapro** [comment]
Here we go;
```
Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry env info

Virtualenv
Python:         3.7.6
Implementation: CPython
Path:           NA

System
Platform: darwin
OS:       posix
Python:   /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7

Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry debug

Poetry
Version: 1.0.2
Python:  3.7.6

Virtualenv
Python:         3.7.6
Implementation: CPython
Path:           NA

System
Platform: darwin
OS:       posix
Python:   /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7
```

**Comment by prabapro** [comment]
Hey @iamkroot , here's an update:

Installed `virtualenv` & reinstalled the trakt-scrobbler.  Now the log is as below;

![image](https://user-images.githubusercontent.com/32594937/72220203-5febf880-3574-11ea-947f-d87ae7217c29.png)

Env info & debug output
![image](https://user-images.githubusercontent.com/32594937/72220242-accfcf00-3574-11ea-826d-acda13e9ae44.png)

**Comment by iamkroot** [comment]
What does the trakt scrobbler log file say?

**Comment by prabapro** [comment]
Here you go;
![image](https://user-images.githubusercontent.com/32594937/72220582-fbcb3380-3577-11ea-9e68-33fdffcc889c.png)

**Comment by iamkroot** [comment]
Hmm no dice. Does your LaunchControl.app provide any additional insights? Try `launchctl start com.github.iamkroot.trakt-scrobbler`

**Comment by prabapro** [comment]
Nothing happening when I do `launchctl start com.github.iamkroot.trakt-scrobbler` 

After installing the `vertualenv`, there is no error in my LaunchControl App.
![image](https://user-images.githubusercontent.com/32594937/72220699-5dd86880-3579-11ea-9463-8682e4317521.png)
When I trying to load the script, getting the same error log.

![image](https://user-images.githubusercontent.com/32594937/72220726-abed6c00-3579-11ea-97d5-0c9b39b5c0d0.png)

**Comment by iamkroot** [comment]
@prabapro Try loading the service first, then run the previous command. Maybe that will help

**Comment by prabapro** [comment]
> @prabapro Try loading the service first, then run the previous command. Maybe that will help

No luck bro :(

**Comment by iamkroot** [comment]
See if running `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python main.py` works inside `/Users/codechilli/Library/trakt-scrobbler` folder.

**Comment by prabapro** [comment]
> See if running `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python main.py` works inside `/Users/codechilli/Library/trakt-scrobbler` folder.

@iamkroot  I'm sorry, I don't know how to do this. If you don't mind, could you kindly let me know the steps to be follow please. Appreciate it.

**Comment by iamkroot** [comment]
> @iamkroot I'm sorry, I don't know how to do this. If you don't mind, could you kindly let me know the steps to be follow please. Appreciate it.

Sorry I should've explained better. You just need to:
1. `cd` to `/Users/codechilli/Library/trakt-scrobbler`
2. run `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python main.py` command directly. (See scrobbler logs to confirm it works)

**Comment by GMFalka** [comment]
I'm monitoring this issue, because I have the same problem on my new Ubuntu install. :)

**Comment by iamkroot** [comment]
> I'm monitoring this issue, because I have the same problem on my new Ubuntu install. :)

Might be a very different issue @GMFalka. Considering this is one is mostly due to an incorrect MacOS service file. Better to create a new issue in case you're unable to fix it.

**Comment by prabapro** [comment]
> > @iamkroot I'm sorry, I don't know how to do this. If you don't mind, could you kindly let me know the steps to be follow please. Appreciate it.
> 
> Sorry I should've explained better. You just need to:
> 
> 1. `cd` to `/Users/codechilli/Library/trakt-scrobbler`
> 2. run `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python main.py` command directly. (See scrobbler logs to confirm it works)

No need to say sorry mate. You are doing such an amazing job. Thank you so much for that. :)

Btw, here is what I'm getting;
![image](https://user-images.githubusercontent.com/32594937/72223096-84a29900-3591-11ea-814a-8d8a66094dec.png)
![image](https://user-images.githubusercontent.com/32594937/72223111-a3a12b00-3591-11ea-8861-dc269b3a4002.png)

**Comment by prabapro** [comment]
> > > @iamkroot I'm sorry, I don't know how to do this. If you don't mind, could you kindly let me know the steps to be follow please. Appreciate it.
> > 
> > 
> > Sorry I should've explained better. You just need to:
> > 
> > 1. `cd` to `/Users/codechilli/Library/trakt-scrobbler`
> > 2. run `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python main.py` command directly. (See scrobbler logs to confirm it works)
> 
> @iamkroot No need to say sorry mate. You are doing such an amazing job. Thank you so much for that. :)
> 
> Btw, here is what I'm getting;
> ![image](https://user-images.githubusercontent.com/32594937/72223096-84a29900-3591-11ea-814a-8d8a66094dec.png)
> ![image](https://user-images.githubusercontent.com/32594937/72223111-a3a12b00-3591-11ea-8861-dc269b3a4002.png)

@iamkroot  Tried this:   `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python3 main.py` (Used `python3` instead of `python`).
![image](https://user-images.githubusercontent.com/32594937/72223229-96387080-3592-11ea-9e9e-e66d4bf6f0d1.png)

**Comment by iamkroot** [comment]
> ![image](https://user-images.githubusercontent.com/32594937/72223111-a3a12b00-3591-11ea-8861-dc269b3a4002.png)

Aha! There we go. Looks like there's no `python` executable inside the virtualenv.
I'll push a fix for this soon. Till then, if you want to test, open up `~/Library/LaunchAgents/trakt-scrobbler.plist` file and in replace `python` with `python3` and save. Then load and start the service again and it should work.

**Comment by prabapro** [comment]
> > ![image](https://user-images.githubusercontent.com/32594937/72223111-a3a12b00-3591-11ea-8861-dc269b3a4002.png)
> 
> Aha! There we go. Looks like there's no `python` executable inside the virtualenv.
> I'll push a fix for this soon. Till then, if you want to test, open up `~/Library/LaunchAgents/trakt-scrobbler.plist` file and replace `python` with `python3` and save. Then load and start the service again and it should work.

@iamkroot Awesome, let me try this & come back to you asap. Thanks heaps.

**Comment by prabapro** [comment]
> Aha! There we go. Looks like there's no `python` executable inside the virtualenv.
> I'll push a fix for this soon. Till then, if you want to test, open up `~/Library/LaunchAgents/trakt-scrobbler.plist` file and replace `python` with `python3` and save. Then load and start the service again and it should work.

Hey @iamkroot ,

Modified the `trakt-scrobbler.plist` with `python3` & reboot the mac. But still getting the same error as before. Nothing is logged in  `/Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_scrobbler.log` too.

However, when I try to run `/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python3 main.py` inside `/Users/codechilli/Library/trakt-scrobbler` folder got below response.;
![image](https://user-images.githubusercontent.com/32594937/72224138-c97ffd00-359c-11ea-92fc-5393a78f9a94.png)
And nothing logged in `trakt_scrobbler.log`

**Comment by iamkroot** [comment]
Hmm sorry, I jumped the gun a bit. Two things:
1. `cd` to `/Users/codechilli/Library/trakt-scrobbler` and give output of `poetry env info -p`
2. Paste the contents of `~/Library/LaunchAgents/trakt-scrobbler.plist` please.

**Comment by prabapro** [comment]
> Hmm sorry, I jumped the gun a bit. Two things:
> 
> 1. `cd` to `/Users/codechilli/Library/trakt-scrobbler` and give output of `poetry env info -p`
> 2. Paste the contents of `~/Library/LaunchAgents/trakt-scrobbler.plist` please.

Hey @iamkroot , no worries. Here is the output;

![image](https://user-images.githubusercontent.com/32594937/72233135-6e292b80-35eb-11ea-92af-01915a7f5b21.png)

Content of `trakt-scrobbler.plist`
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"
    "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.github.iamkroot.trakt-scrobbler</string>
    <key>ProgramArguments</key>
    <array>
        <string>/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3</string>
        <string>main.py</string>
    </array>
    <key>WorkingDirectory</key>
    <string>/Users/codechilli/Library/trakt-scrobbler</string>
    <key>RunAtLoad</key>
    <true />
    <key>LaunchOnlyOnce</key>
    <true />
    <key>KeepAlive</key>
    <true />
</dict>
</plist>
```

**Comment by iamkroot** [comment]
Could you check if `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/` exists and contains a `python3` file? Also, try running `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3 main.py` from inside `/Users/codechilli/Library/trakt-scrobbler`

**Comment by prabapro** [comment]
> Could you check if `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/` exists and contains a `python3` file? Also, try running `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3 main.py` from inside `/Users/codechilli/Library/trakt-scrobbler`

1. Yes - `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/` exists. Files inside are shown below; (inside there is a shortcut for `python3`).
![image](https://user-images.githubusercontent.com/32594937/72265472-942dea80-3642-11ea-8f80-2ef8450f0c5f.png)

2. Ran `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3 main.py` inside `/Users/codechilli/Library/trakt-scrobbler` folder & got below response;
![image](https://user-images.githubusercontent.com/32594937/72265903-4bc2fc80-3643-11ea-912b-3b4018c06bd3.png)

**Comment by iamkroot** [comment]
>     ![image](https://user-images.githubusercontent.com/32594937/72265903-4bc2fc80-3643-11ea-912b-3b4018c06bd3.png)

Looks like the dependencies haven't been installed properly for some reason. Run the install script again (from where you downloaded the project)

**Comment by prabapro** [comment]
> > ```
> > ![image](https://user-images.githubusercontent.com/32594937/72265903-4bc2fc80-3643-11ea-912b-3b4018c06bd3.png)
> > ```
> 
> Looks like the dependencies haven't been installed properly for some reason. Run the install script again (from where you downloaded the project)

Hey @iamkroot , installed the script once again & it's still same. 

```
Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry env info

Virtualenv
Python:         3.7.6
Implementation: CPython
Path:           /Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7
Valid:          True

System
Platform: darwin
OS:       posix
Python:   /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7
Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry debug

Poetry
Version: 1.0.2
Python:  3.7.6

Virtualenv
Python:         3.7.6
Implementation: CPython
Path:           /Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7
Valid:          True

System
Platform: darwin
OS:       posix
Python:   /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7
Prabas-MacBook-Air:trakt-scrobbler codechilli$ launchctl list | grep com.github.iamkroot.trakt-scrobbler
Prabas-MacBook-Air:trakt-scrobbler codechilli$ /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/bin/python3 main.py
Traceback (most recent call last):
  File "main.py", line 10, in <module>
    from player_monitors.monitor import Monitor
  File "/Users/codechilli/Library/trakt-scrobbler/player_monitors/monitor.py", line 3, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
Prabas-MacBook-Air:trakt-scrobbler codechilli$ poetry env info -p
/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3PPrabaPraPrabPrPrabaPrabPraPraPrPrPPrPPPPPPPPPPPPPPrabas-MacBook-Air:trakt-scrobbler codechilli$ 
```

**Comment by iamkroot** [comment]
What's the output when you run `poetry install` inside the installation directory (`/Users/codechilli/Library/trakt-scrobbler`)?

**Comment by prabapro** [comment]
@iamkroot Here we go;
![image](https://user-images.githubusercontent.com/32594937/72351687-d587ce00-3706-11ea-80ee-1a661b2539f1.png)

Btw, wondering if we need to install `requests` module.... Pls advice. Thanks

**Comment by iamkroot** [comment]
@prabapro try the install script now, reboot and check logs

**Comment by prabapro** [comment]
> @prabapro try the install script now, reboot and check logs

Hi @iamkroot, tried as instructed. But still same, no new entries in the log. 

Response for running `/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3 main.py` inside `/Users/codechilli/Library/trakt-scrobbler`
![image](https://user-images.githubusercontent.com/32594937/72361196-1556b180-3717-11ea-891a-b9e40077ef3d.png)

**Comment by iamkroot** [comment]
@prabapro Let's give it one more try, in the installation dir, run `poetry run python main.py` and check log to see if it works. (No need to reboot)

**Comment by prabapro** [comment]
@iamkroot sure thing :) Here's the log (For the first time, I'm seeing something in the log :)

```
2020-01-14 21:26:29,727 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_cache.json doesn't exist.
2020-01-14 21:26:29,727 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_token.json doesn't exist.
2020-01-14 21:26:29,727 - INFO - MainThread - trakt_interface - Access token not found in config. Initiating device authentication.
2020-01-14 21:26:30,565 - INFO - MainThread - trakt_interface - Verification URL: https://trakt.tv/activate
2020-01-14 21:26:30,565 - INFO - MainThread - trakt_interface - User Code: 49361619
2020-01-14 21:26:31,866 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:26:37,805 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:26:43,520 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:26:49,224 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:26:54,858 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:27:00,635 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:27:06,332 - DEBUG - MainThread - trakt_interface - Waiting for user to authorize the app.
2020-01-14 21:27:12,435 - INFO - MainThread - trakt_interface - Device auth successful.

2020-01-14 22:19:14,882 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_cache.json doesn't exist.
2020-01-14 22:19:14,884 - INFO - MainThread - scrobbler - Started scrobbler thread.
2020-01-14 22:19:14,884 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/watched_cache.json doesn't exist.
2020-01-14 22:19:14,892 - INFO - MainThread - monitor - Started monitor for mpv
2020-01-14 22:19:14,893 - INFO - mpv - mpv - Unable to connect to MPV. Check ipc path.
2020-01-14 22:19:14,897 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/plex_token.json doesn't exist.
2020-01-14 22:19:16,095 - ERROR - MainThread - main - Unhandled exception
Traceback (most recent call last):
  File "main.py", line 78, in <module>
    main()
  File "main.py", line 73, in main
    mon = Mon(scrobble_queue)
  File "/Users/codechilli/Library/trakt-scrobbler/player_monitors/plex.py", line 45, in __init__
    self.token = get_token(**plex_conf)
  File "/Users/codechilli/Library/trakt-scrobbler/player_monitors/plex.py", line 32, in get_token
    write_json({"token": token}, PLEX_TOKEN_PATH)
  File "/Users/codechilli/Library/trakt-scrobbler/utils.py", line 36, in write_json
    json.dump(data, f, indent=4)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type Response is not JSON serializable
```

**Comment by iamkroot** [comment]
Ah progress :confetti_ball: 

I'm not exactly sure why you're getting that `TypeError` (seems to be working fine in my system), but we might have a solution for your installation errors.

In your plist file, replace 
```xml
<key>ProgramArguments</key>
    <array>
        <string>/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3</string>
        <string>main.py</string>
    </array>
```
with
```xml
<key>ProgramArguments</key>
    <array>
        <string>poetry</string>
        <string>run</string>
        <string>python</string>
        <string>main.py</string>
    </array>
```

**Comment by iamkroot** [comment]
Closed by mistake, sorry.

**Comment by prabapro** [comment]
> Closed by mistake, sorry.

haha, no worries. Let me do above & come back to you in few mins. Thanks mate

**Comment by prabapro** [comment]
> Ah progress 🎊
> 
> I'm not exactly sure why you're getting that `TypeError` (seems to be working fine in my system), but we might have a solution for your installation errors.
> 
> In your plist file, replace
> 
> ```
> <key>ProgramArguments</key>
>     <array>
>         <string>/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3</string>
>         <string>main.py</string>
>     </array>
> ```
> 
> with
> 
> ```
> <key>ProgramArguments</key>
>     <array>
>         <string>poetry</string>
>         <string>run</string>
>         <string>python</string>
>         <string>main.py</string>
>     </array>
> ```
> 
> And then load the service using `launchctl` and reboot. Then check logs to see if it worked.

Hey @iamkroot, seems we are on track :)

**Task 1: Modified plist file & loaded (rebooted too). below is the log;**

![image](https://user-images.githubusercontent.com/32594937/72368670-8cdf0d80-3724-11ea-96a2-8d50ca459d91.png)

![image](https://user-images.githubusercontent.com/32594937/72368735-a5e7be80-3724-11ea-805f-e7e4a9be640f.png)

**Task 2: Tried `poetry run python main.py`**

Scrobbler started for Plex. 👍  VLC got failed. Below is the log & vlc settings.

```
2020-01-14 22:32:36,834 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_cache.json doesn't exist.
2020-01-14 22:32:36,835 - INFO - MainThread - scrobbler - Started scrobbler thread.
2020-01-14 22:32:36,840 - INFO - MainThread - monitor - Started monitor for mpv
2020-01-14 22:32:36,841 - INFO - mpv - mpv - Unable to connect to MPV. Check ipc path.
2020-01-14 22:32:36,844 - WARNING - MainThread - utils - Invalid json in /Users/codechilli/Library/Application Support/trakt-scrobbler/plex_token.json.
2020-01-14 22:32:38,016 - ERROR - MainThread - main - Unhandled exception
Traceback (most recent call last):
  File "main.py", line 78, in <module>
    main()
  File "main.py", line 73, in main
    mon = Mon(scrobble_queue)
  File "/Users/codechilli/Library/trakt-scrobbler/player_monitors/plex.py", line 45, in __init__
    self.token = get_token(**plex_conf)
  File "/Users/codechilli/Library/trakt-scrobbler/player_monitors/plex.py", line 32, in get_token
    write_json({"token": token}, PLEX_TOKEN_PATH)
  File "/Users/codechilli/Library/trakt-scrobbler/utils.py", line 36, in write_json
    json.dump(data, f, indent=4)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type Response is not JSON serializable
2020-01-14 23:12:41,748 - DEBUG - MainThread - utils - /Users/codechilli/Library/Application Support/trakt-scrobbler/trakt_cache.json doesn't exist.
2020-01-14 23:12:41,750 - INFO - MainThread - scrobbler - Started scrobbler thread.
2020-01-14 23:12:41,759 - WARNING - MainThread - utils - Invalid json in /Users/codechilli/Library/Application Support/trakt-scrobbler/plex_token.json.
2020-01-14 23:12:43,222 - INFO - MainThread - monitor - Started monitor for plex
2020-01-14 23:12:43,232 - INFO - MainThread - monitor - Started monitor for vlc
2020-01-14 23:12:43,443 - INFO - vlc - monitor - Unable to connect to vlc. Ensure that the web interface is running.
2020-01-14 23:13:44,715 - DEBUG - scrobbler - trakt_interface - Searching cache.
2020-01-14 23:13:44,715 - DEBUG - scrobbler - trakt_interface - Searching trakt.
2020-01-14 23:14:30,928 - DEBUG - scrobbler - trakt_interface - Trakt ID: 414606
2020-01-14 23:14:30,930 - DEBUG - scrobbler - trakt_interface - {'movie': {'ids': {'trakt': 414606}}}
2020-01-14 23:15:53,031 - INFO - scrobbler - scrobbler - Scrobble start successful.
2020-01-14 23:15:55,701 - DEBUG - scrobbler - trakt_interface - Searching cache.
2020-01-14 23:15:55,701 - DEBUG - scrobbler - trakt_interface - {'movie': {'ids': {'trakt': 414606}}}
2020-01-14 23:16:13,205 - INFO - scrobbler - scrobbler - Scrobble stop successful.
2020-01-14 23:16:15,587 - DEBUG - scrobbler - trakt_interface - Searching cache.
2020-01-14 23:16:15,587 - DEBUG - scrobbler - trakt_interface - Searching trakt.
2020-01-14 23:16:44,690 - DEBUG - scrobbler - trakt_interface - Trakt ID: 413326
2020-01-14 23:16:44,691 - DEBUG - scrobbler - trakt_interface - {'movie': {'ids': {'trakt': 413326}}}
2020-01-14 23:16:59,569 - INFO - scrobbler - scrobbler - Scrobble pause successful.
2020-01-14 23:16:59,569 - DEBUG - scrobbler - trakt_interface - Searching cache.
2020-01-14 23:16:59,570 - DEBUG - scrobbler - trakt_interface - {'movie': {'ids': {'trakt': 413326}}}
2020-01-14 23:17:13,040 - INFO - scrobbler - scrobbler - Scrobble start successful.
2020-01-14 23:19:55,382 - DEBUG - scrobbler - trakt_interface - Searching cache.
2020-01-14 23:19:55,383 - DEBUG - scrobbler - trakt_interface - {'movie': {'ids': {'trakt': 413326}}}
2020-01-14 23:20:07,163 - INFO - scrobbler - scrobbler - Scrobble stop successful.
```
![image](https://user-images.githubusercontent.com/32594937/72369233-a896e380-3725-11ea-973f-33bc281572d2.png)

![image](https://user-images.githubusercontent.com/32594937/72369263-b6e4ff80-3725-11ea-82a4-4f0f31fd0821.png)

**Comment by prabapro** [comment]
Hey @iamkroot , happy news 🥳 both the VLC & Plex are getting scrobbled now.

Only issue we have right now is the plist file. 

For the time being I'v set up an alias in `bash_profile` to run  `cd /Users/codechilli/Library/trakt-scrobbler && poetry run python main.py`

**Comment by mastaka** [comment]
> Created an Automator task as below (for both VLC & Plex);
>Now, whenever I start either VLC or Plex, the scrobbler starts automatically :)

Hi! I can’t understand how to make it so that when starting the VLC the script just runs, instead of starting the service or application created in the Automator? 
How do you save this code for Automator? 

P.S. Sorry, if my questions seem stupid, I am poorly versed in the Automator, and even more so in programming.

**Comment by mastaka** [comment]
> I'm not exactly sure why you're getting that `TypeError` (seems to be working fine in my system), but we might have a solution for your installation errors.
> 
> In your plist file, replace
> 
> ```
> <key>ProgramArguments</key>
>     <array>
>         <string>/Users/codechilli/Library/Caches/pypoetry/virtualenvs/trakt-scrobbler-SsqFgJVg-py3.7/bin/python3</string>
>         <string>main.py</string>
>     </array>
> ```
> 
> with
> 
> ```
> <key>ProgramArguments</key>
>     <array>
>         <string>poetry</string>
>         <string>run</string>
>         <string>python</string>
>         <string>main.py</string>
>     </array>
> ```
> 
> And then load the service using `launchctl` and reboot. Then check logs to see if it worked.

Hi @iamkroot ! I have almost the same problems as @prabapro 
Service after restart not working in both versions in `launchctl`, but manual `poetry run python main.py` from trakt-scrobbler  directory is working and VLC scrobbled.

**Comment by prabapro** [comment]
> > Created an Automator task as below (for both VLC & Plex);
> > Now, whenever I start either VLC or Plex, the scrobbler starts automatically :)
> 
> Hi! I can’t understand how to make it so that when starting the VLC the script just runs, instead of starting the service or application created in the Automator?
> How do you save this code for Automator?
> 
> P.S. Sorry, if my questions seem stupid, I am poorly versed in the Automator, and even more so in programming.

Hi @mastaka, as you mentioned [above](https://github.com/iamkroot/trakt-scrobbler/issues/42#issuecomment-574609862) running `poetry run python main.py` in the installation directory is the only option I do have right now. I'm also new to Automator &  managed to come up with this after reading a few stuff from the Stakeworkflow. I don't know the pros/cons of this method but @iamkroot would be able to enlighten us on this. Below are the steps I used to achieve this, FYI.

So the idea is to setting up a workflow where once we open the VLC, automator will trigger the `poetry run python main.py` automatically.

1. Launch the Automator app > `Workflow` > Choose
2. In the search bar, type `launch`
3. Double click on `Launch Application`
4. Select `VLC` in the dropdown (right hand pane - you might have to click on `other` & choose `vlc.app`)
5. Again on the left pane, search `run` & double click on `Run Shell Script`
6. Make sure the shell selected is `/bin/bash`
7. Type your terminal command here. First we need to `cd` to the installation directory followed by `poetry run python main.py` command. But `poetry` doesn't work inside the automator. So we have to replace `poetry` with it's path. To do that, open up a Terminal window and run `which poetry` - in my case the path is `/usr/local/bin/poetry`
- So here is the format: `cd <installation dir> && <poetry path> run python main.py`.
- For me it's like `cd /Users/codechilli/Library/trakt-scrobbler && /usr/local/bin/poetry run python main.py`

![image](https://user-images.githubusercontent.com/32594937/72433702-f06a4900-37bf-11ea-85f5-da936d7e4368.png)

8. Now go to`File` > `Save`
9. Once saved, launch the VLC & play a video & check the `trakt_scrobbler.log` for the status.

**Comment by mastaka** [comment]
@prabapro thank you very match for such a detailed tutorial!

I probably didn’t put it right, sorry. I created my workflow after your first comment about the Automator, but the problem is that it only works inside the Automator. Or if I save it as an application or service and run it. 

But if, after saving the workflow , I just start the VLC, then scrobbling does not occur.

![Run VLC and Scrobble Workflow 2020-01-15 14-32-17](https://user-images.githubusercontent.com/10599155/72437817-e20e3400-37a3-11ea-962e-8eddddde1594.png)

**Comment by prabapro** [comment]
Hi @mastaka , you're most welcome :)

I couldn't figure out the reason for your issue. See if this works: Double click on the saved file (`xxxx.app`) & see if vlc opens. And then you can check the log file at `/Users/mstk/Library/Application Support/trakt-scrobbler/trakt_scrobbler.log`.

**Comment by mastaka** [comment]
@prabapro That is what I had in mind. Since you have not indicated anywhere that you are launching the application saved from Automator (yes, this also works for me). 

But from your words, I thought that you just start the VLC as usual, and after that the script for scrobbling is automatically launched — therefore I couldn’t understand how you do it :D 

Thanks for the detailed answers! We just did not understand each other :)

**Comment by prabapro** [comment]
> @prabapro That is what I had in mind. Since you have not indicated anywhere that you are launching the application saved from Automator (yes, this also works for me).
> 
> But from your words, I thought that you just start the VLC as usual, and after that the script for scrobbling is automatically launched — therefore I couldn’t understand how you do it :D
> 
> Thanks for the detailed answers! We just did not understand each other :)

Hey @mastaka , no worries :)

I think we have to run the app created by the Automator only once (after each reboot). Thereafter opening the VLC normally would trigger the scrobbler. No need to launch the created app every time. I'm not sure about the reason behind this :)

**Comment by mastaka** [comment]
@prabapro Although this is a “crutch”, but in this case the problem can be solved as follows. 
Create an application in the machine with a script only and add it to autoload.
![Users   Groups 2020-01-15 15-45-29](https://user-images.githubusercontent.com/10599155/72443943-7336d800-37af-11ea-85b1-ae3d4ca22e38.png)

After the restart, the script will be executed, and scrobbling will work, and if necessary, the script can be interrupted through the gear.
![Mozilla Firefox 2020-01-15 15-48-50(1)](https://user-images.githubusercontent.com/10599155/72443993-8c3f8900-37af-11ea-8e4e-d11857c34ad9.png)

**Comment by prabapro** [comment]
> @prabapro Although this is a “crutch”, but in this case the problem can be solved as follows.
> Create an application in the machine with a script only and add it to autoload.
> ![Users Groups 2020-01-15 15-45-29](https://user-images.githubusercontent.com/10599155/72443943-7336d800-37af-11ea-85b1-ae3d4ca22e38.png)
> 
> After the restart, the script will be executed, and scrobbling will work, and if necessary, the script can be interrupted through the gear.
> ![Mozilla Firefox 2020-01-15 15-48-50(1)](https://user-images.githubusercontent.com/10599155/72443993-8c3f8900-37af-11ea-8e4e-d11857c34ad9.png)

@mastaka wow, this is awesome. Let me follow this. Thanks heaps :)

**Comment by iamkroot** [comment]
@mastaka Great work with that!

**Comment by iamkroot** [comment]
@prabapro One more thing you can try, is in the plist file, replace `poetry` with `/usr/local/bin/poetry` and then reload the service. Maybe that might work. (Check logs to confirm)

**Comment by mastaka** [comment]
> One more thing you can try, is in the plist file, replace `poetry` with `/usr/local/bin/poetry` and then reload the service. Maybe that might work. (Check logs to confirm)

Hi @iamkroot ! Bingo! This is working :) Thank you!
![trakt-scrobbler plist 2020-01-16 11-43-19](https://user-images.githubusercontent.com/10599155/72518651-180bf080-3856-11ea-958b-6d18351b5adf.png)
![trakt_scrobbler log 2020-01-16 11-50-09](https://user-images.githubusercontent.com/10599155/72518835-6de09880-3856-11ea-86e2-2e8512f388bf.png)

**Comment by iamkroot** [comment]
Woohoo! I'll push an update ASAP with the updated code.

**Comment by prabapro** [comment]
> @prabapro One more thing you can try, is in the plist file, replace `poetry` with `/usr/local/bin/poetry` and then reload the service. Maybe that might work. (Check logs to confirm)

Hey @iamkroot , 

After replacing `poetry` with `/usr/local/bin/poetry` & loaded & rebooted, the system log is as below;
```
Jan 16 20:56:28 Prabas-MacBook-Air com.apple.xpc.launchd[1] (com.github.iamkroot.trakt-scrobbler): This service is defined to be constantly running and is inherently inefficient.
Jan 16 20:56:31 Prabas-MacBook-Air com.apple.xpc.launchd[1] (com.github.iamkroot.trakt-scrobbler[1351]): Service exited with abnormal code: 1
```
Here's the content of my `trakt-scrobbler.plist` file;
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"
    "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.github.iamkroot.trakt-scrobbler</string>
    <key>ProgramArguments</key>
    <array>
        <string>/usr/local/bin/poetry</string>
        <string>run</string>
        <string>python</string>
        <string>main.py</string>
    </array>
    <key>WorkingDirectory</key>
    <string>/Users/codechilli/Library/trakt-scrobbler</string>
    <key>RunAtLoad</key>
    <true />
    <key>LaunchOnlyOnce</key>
    <true />
    <key>KeepAlive</key>
    <true />
</dict>
</plist>
```
PS: @mastaka's [plist file](https://github.com/iamkroot/trakt-scrobbler/issues/42#issuecomment-575095472) is little different from mine. Could that be a reason for this? I'm running on `Catalina OS X 10.15.2`

**Comment by prabapro** [comment]
> > @prabapro One more thing you can try, is in the plist file, replace `poetry` with `/usr/local/bin/poetry` and then reload the service. Maybe that might work. (Check logs to confirm)
> 
> Hey @iamkroot ,
> 
> After replacing `poetry` with `/usr/local/bin/poetry` & loaded & rebooted, the system log is as below;
> 
> ```
> Jan 16 20:56:28 Prabas-MacBook-Air com.apple.xpc.launchd[1] (com.github.iamkroot.trakt-scrobbler): This service is defined to be constantly running and is inherently inefficient.
> Jan 16 20:56:31 Prabas-MacBook-Air com.apple.xpc.launchd[1] (com.github.iamkroot.trakt-scrobbler[1351]): Service exited with abnormal code: 1
> ```
> 
> Here's the content of my `trakt-scrobbler.plist` file;
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"
>     "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
> <plist version="1.0">
> <dict>
>     <key>Label</key>
>     <string>com.github.iamkroot.trakt-scrobbler</string>
>     <key>ProgramArguments</key>
>     <array>
>         <string>/usr/local/bin/poetry</string>
>         <string>run</string>
>         <string>python</string>
>         <string>main.py</string>
>     </array>
>     <key>WorkingDirectory</key>
>     <string>/Users/codechilli/Library/trakt-scrobbler</string>
>     <key>RunAtLoad</key>
>     <true />
>     <key>LaunchOnlyOnce</key>
>     <true />
>     <key>KeepAlive</key>
>     <true />
> </dict>
> </plist>
> ```
> 
> PS: @mastaka's [plist file](https://github.com/iamkroot/trakt-scrobbler/issues/42#issuecomment-575095472) is little different from mine. Could that be a reason for this? I'm running on `Catalina OS X 10.15.2`

Hey @iamkroot & @mastaka, happy news 👍  Managed to get this worked successfully. Ran `poetry install` inside the installation dir, viola 🎊 Thanks heaps boys!!!
